## **Getting Started with Python**

### **1. Hello World Program!**

In [1]:
# printing "Hello World" because its the rule!
print("Hello World!")

Hello World!


**Exploring More!**

In [2]:
print(20+50)

70


In [3]:
print('20+50')

20+50


> Because in `'' or ""` python recognize them as sentence (string). That's why we got as it is value in second time we run the cell `print('20+50')`

In [4]:
print(30>5)

True


In [5]:
print(6<5)

False


> Here, `30` is greater than 5 which is `True` thats why we got True as an output.

### **2. Variables and Data Types**

In [1]:
# creating a python variable is super easy, all we have to do is:
a = 15

**Rules To create a variable in python**

In [2]:
# variable name should not starts with number
1a = 15

SyntaxError: invalid decimal literal (2269455664.py, line 2)

In [3]:
a1 = 15

In [4]:
# variable name should not have space
a 2 = 10

SyntaxError: invalid syntax. Perhaps you forgot a comma? (746718547.py, line 2)

In [5]:
a_2 = 15

**Memory address alocation in python**

In [19]:
a = 10

In [20]:
print("Address of a = ",id(a))

Address of a =  1857617527312


In [21]:
b = a

In [22]:
print("Address of a = ",id(b))

Address of a =  1857617527312


In [23]:
print("Address of b = ",id(b))

Address of b =  1857617527312


In [24]:
print("Is addresses of both variables are same?",id(a)==id(b))

Is addresses of both variables are same? True


In [25]:
# lets change the value of variable a
a = 'Hello'
print(id(a))

1857698929456


**What I did here?**
> As you can see that the new value assigned to a different address which means that the `variable a` still holds the prev data in address = `1857617527312` and new data is stored in new address = `1857698929456`. Which means that this data type cannot be changed therefore we say that this is an **immutable variable**.

### **Mutable & Immutable Data Type**

- **Mutable Data Types (can be modified)**
    - List
    - Dictionary
    - byte array
- **Immutable Data Types (can't be modified)**
    - Int
    - Float
    - Complex
    - String
    - Tuple
    - Set

### **Data Types in Python**
In python we have following data types :
- Numeric
    - Integers 
    - Float
    - Complex Numbers
- Sequence Types
    - String
    - List
    - Tuple
- Dictionary
- Set

#### **Numeric Types**

In [2]:
a = 5
b = 5.5
c = 3.136468436894368684378567897698372689

In [3]:
type(a)

int

In [4]:
type(b)

float

In [5]:
type(c)

float

In [6]:
d = 5j

In [7]:
type(d)

complex

#### **Sequence Data Types**
1. **String Data Types**

In [9]:
varA = 'Hello this is an example of string data type (basically sentences or words)'

In [10]:
# To check type of a variable we use `type()` func
print('Type of varA = ',type(varA))

Type of varA =  <class 'str'>


> **One noteable thing is here that in python we don't have any data types , we have `data classes` instead.**

**Operation on strings**

In [12]:
# string concatination / addition of two or more strings
string1 = 'Hello baby '
string2 = 'How do you do'
print(string1 + string2)

Hello baby How do you do


In [13]:
# another way
print('Hello baby '+'How do you do')

Hello baby How do you do


In [14]:
# capitalization
print(string1.capitalize())

Hello baby 


In [15]:
# uppercase
print(string2.upper())

HOW DO YOU DO


In [16]:
# lower
print(string1.lower())

hello baby 


In [29]:
#  acessing a letter of a string through indexing
string1[0],string1[1],string1[2],string1[3],string1[4],string1[5],string1[6],string1[7],string1[8],string1[9],string1[10],string1[11]

IndexError: string index out of range

> It means in `string1` we only have 12 characters (whitespace included).

In [30]:
string1[0],string1[1],string1[2],string1[3],string1[4],string1[5],string1[6],string1[7],string1[8],string1[9],string1[10]

('H', 'e', 'l', 'l', 'o', ' ', 'b', 'a', 'b', 'y', ' ')

In [43]:
string1[0:9] # slicing string through a start index and end index

'Hello bab'

In [44]:
string1[0:7:2] # slicing strig

'Hlob'

2. **List**

In [45]:
a = [1,2.2,"panda"]

In [46]:
type(a)

list

In [47]:
#  accessing a list value throug idx
a[0]

1

In [48]:
a[1]

2.2

In [50]:
a[2]

'panda'

In [51]:
a[3]

IndexError: list index out of range

In [52]:
# updating a list element
a[0] = 'hello'

In [53]:
a

['hello', 2.2, 'panda']

2. **Tuple**

In [54]:
tuple1 = (1,2,3,0.00000000004,5)

In [55]:
type(tuple1)

tuple

In [60]:
tuple2 = 1,2,3,0.0000000000004,5 # another way of defining a tuple

In [59]:
type(tuple2)

tuple

In [61]:
# accessing a tuple element through idx
tuple1[0]

1

In [62]:
tuple1[2]

3

In [64]:
tuple1[3]

4e-11

In [65]:
# lets try to change a value of tuple
tuple1[0] =10

TypeError: 'tuple' object does not support item assignment

#### **Dictionaries**
A data type of key-value pair

In [66]:
dict1 = {
    "name":['vivek','snehu'],
    "branch":["cse","mech"],
    "class_no":[211,208]
}

In [67]:
dict1["name"]

['vivek', 'snehu']

In [68]:
dict1["branch"]

['cse', 'mech']

In [69]:
dict1["class_no"]

[211, 208]

> good old days :)

In [73]:
# modifying a value in dictionary
dict1["name"] = 'nobody'
dict1["branch"] = 'left the college'
dict1["class_no"] = 'left the college'

In [74]:
dict1

{'name': 'nobody',
 'branch': 'left the college',
 'class_no': 'left the college'}

In [75]:
# adding a new-key value pair
dict1["what_happend"] = 'end'
dict1['betrayal'] = True

In [76]:
dict1

{'name': 'nobody',
 'branch': 'left the college',
 'class_no': 'left the college',
 'what_happend': 'end',
 'betrayal': True}

> sad ending :(

#### **Set**

In [77]:
set1 = {'anshu','siru','dhanashree'}

In [80]:
type(set1)

set

In [78]:
set2 =  {'anshu','siru','dhanashree','anshu'}

In [79]:
set2

{'anshu', 'dhanashree', 'siru'}

> set removes the duplicates just as I do :)

### **Getting User Input & Type Casting**

In [1]:
var_a = input("Enter the value = ")

In [3]:
var_a

'5'

In [2]:
# lets check the default input type
type(var_a)

str

> Even through we have given a number as an input python still recognize this as `str` type

**Type Casting or Type Conversation**

In [4]:
# converting into a number
converted_var_a = int(var_a)

In [5]:
type(converted_var_a)

int

In [6]:
# but it has some limitations, like
var_b = 'Five'
int(var_b)

ValueError: invalid literal for int() with base 10: 'Five'

In [8]:
var_c = '5'
type(int(var_c))

int

> Values like : 5,6,0.6,100 can be converted to float and int but values like 'Five','Hundered' can't. This logic same goes for converting numbers to int or float.

In [10]:
# converting a value to float
var_a = '5.678'
type(var_a)

str

In [11]:
type(float(var_a))

float

In [16]:
# converting a value to boolean
var_b = 1
type(var_b)

int

In [17]:
type(bool(var_b))

bool

In [18]:
bool(var_b)

True

In [19]:
# converting a str to int
var_c = 'five'
int(var_c)

ValueError: invalid literal for int() with base 10: 'five'

In [22]:
var_c = '1000'
var_c = int(var_c)

In [23]:
type(var_c)

int

> Same goes for float

### **Operators**

1. **Arithmetic Operators**

In [27]:
a,b = 5,10

In [28]:
# addition
a+b

15

In [34]:
# substraction
b-a

5

In [55]:
# mutiplication
a*b

50

In [56]:
# division
b/a

2.0

In [58]:
# modulus
b%3

1

> 10/3 = 3 with 1 as remainder there for we got 1 as answer

In [39]:
# exponent
a**2

25

> It means 5 to power 2 or a^n, In our case it means 5^2 = 5x5 = 25

In [42]:
# floor divison
b//3

3

> As b = 10 => b/3 => 10/3 = 3 + 1 as remainder, Here it removes remainder 1 and returns output 3

In [44]:
# diff btw floor divison and normal divison
print("10/3 with normal divison = ",10/3)
print("10/3 with floor divison = ",10//3)

10/3 with normal divison =  3.3333333333333335
10/3 with floor divison =  3


**Assignment Operators**

In [72]:
x = 5 # same as x is equals to 5

In [73]:
y = 5
y += 5 # y = y + 5
y # here we have y = 5 => y += 5 => y = y+ 5 => y = 5 + 5 => y = 10

10

In [74]:
y = 5
y -= 5 # y = y - 5
y # here we have y = 5 => y -= 5 => y = y - 5 => y = 5 + 5 => y = 0

0

In [75]:
y = 5
y *= 5 # y = y x 5
y # here we have y = 5 => y *= 5 => y = y x 5 => y = 5 x 5 => y = 25

25

In [76]:
y = 5
y /= 5 # y = y / 5
y # here we have y = 5 => y /= 5 => y = y / 5 => y = 5 / 5 => y = 1

1.0

In [78]:
y = 5
y **= 5 # y = y^5
y # here we have y = 5 => y **= 5 => y = y ^ 5 => y = 5 x 5  x 5 x 5 x 5 => y = 3125

3125

In [79]:
y = 5
y //= 5 # y = y // 5
y # here we have y = 5 => y //= 5 => y = y // 5 => y = 5 // 5 => y = 1

1

In [81]:
y = 5
y %= 5 # y = y % 5
y # here we have y = 5 => y %= 5 => y = y % 5 => y = 5 % 5 => y = 0

0

**Comparison Operator**

In [82]:
p,q = 10,20

In [83]:
p>q

False

In [85]:
p<q

True

In [86]:
p==q

False

In [87]:
p<=q

True

In [88]:
p>=q

False

In [89]:
q>p

True

In [90]:
q<p

False

In [91]:
q==p

False

In [92]:
q>=p

True

In [93]:
q<=p

False

In [94]:
q!=p

True

In [95]:
p!=q

True

**Logical Operator**

1. **AND**
> Returns `True` if both statements were `True` and `False` if both were `False` else `False`.

In [2]:
True and False

False

In [3]:
True and True

True

In [4]:
False and True

False

In [5]:
False and False

False

In [14]:
# example 1 : statement 1 is True and statement 2 is also True
3 > 2 and 4 > 3

True

In [16]:
# example 2 : statement 1 is False and statement 2 is True
3 < 2 and 4 > 3

False

In [20]:
# example 3 : statement 1 is True and statement 2 is False
2 < 3 and 3 < 0

False

In [21]:
4 > 5 and 3 > 4 # statement 1 is False and statement 2 is False

False

**OR**
> Returns `False` if both the statement were False

In [22]:
True or True

True

In [23]:
False or True

True

In [24]:
True or False

True

In [25]:
False or False

False

**NOT**
> Returns `True` if False and `False` if True

In [26]:
not True

False

In [27]:
not False

True

**Membership Operator**
> Checks if an element is present in a node or not

In [28]:
sentence = "Hello baby how do you do?"

In [29]:
"Hello" in sentence

True

In [30]:
"Hello" not in sentence

False

In [31]:
"baby" in sentence

True

In [32]:
"babu" in sentence

False

In [33]:
List1 = [1,2,3,4,5]

In [34]:
1 in List1

True

In [35]:
7 in List1

False

**Identity Operator**

In [42]:
person1, person2 = "Vivek", "Subhashree"
person1 is person2 # ?

False

In [43]:
person1 == person2

False

In [44]:
person1 is not person2

True

In [45]:
person1 != person2

True

**Bitwise Operator**

In [2]:
a,b = int(True), int(False)
a,b

(1, 0)

**&**
> bitwise AND

In [3]:
a & b

0

In [50]:
b & a

0

In [5]:
a & a

1

In [6]:
b & b

0

**||**
> bitwise OR

In [7]:
a | b

1

In [8]:
b | a

1

In [9]:
a | b

1

In [11]:
b | b

0